File: src/com/music/read/PlayManager.java
package com.music.read;

import javafx.application.Platform;
import javafx.scene.image.Image;

import java.io.ByteArrayInputStream;

/**
 * Created by xupanpan on 03/11/2017.
 */
public class PlayManager {
    private HomeView homeView;

    private boolean isUserControl;

    public enum PlayType {
        SINGLE, RECYCLE, RANDOM
    }

    private PlayType playType = PlayType.RECYCLE;

    public PlayType getPlayType() {
        return playType;
    }

    public void setPlayType(PlayType playType) {
        this.playType = playType;
    }

    private PlayScheduleListener listener = new PlayScheduleListener() {
        public void onPlaying(final int position) {
            Platform.runLater(new Runnable() {
                public void run() {
                    homeView.playTimeLabel.setText(Utils.getMusicTime(position));
                }
            });
        }
    };


    public PlayManager(HomeView view) {
        this.homeView = view;
        MusicPlayer.getInstans().setPlayScheduleListener(listener);
    }

    public void play(int position) {
        play(DataManager.getInstans().getMP3InfoByPosition(position));
    }


    public void play(MP3Info info) {
        if (info == null) {
            return;
        }

        DataManager.getInstans().setNewPlayPosition(info);
        isUserControl = true;
        playMusic();
    }

    public void playNextAuto() {

        switch (playType) {
            case RANDOM:
                DataManager.getInstans().setPlayNextRadomPosition();
                break;
            case RECYCLE:
                DataManager.getInstans().setPlayNextPosition();
                break;
            case SINGLE:

                break;
        }

        playMusic();
        homeView.scrollToShow();
    }

    public void playNext() {
        isUserControl = true;
        playNextAuto();
    }


    public void playLast() {
        isUserControl = true;

        switch (playType) {
            case RANDOM:
                DataManager.getInstans().setPlayNextRadomPosition();
                break;
            case RECYCLE:
                DataManager.getInstans().setPlayLastPosition();
                break;
            case SINGLE:

                break;
        }
        playMusic();
        homeView.scrollToShow();
    }

    public void stopAndStart() {

        if (MusicPlayer.getInstans().isPlaying()) {
            isUserControl = true;
            MusicPlayer.getInstans().closePlay();

        } else {
            playMusic();
            homeView.setButtonPlay();
            homeView.scrollToShow();
        }
    }


    private PlayStateListener mPlayStateListener = new PlayStateListener() {

        public void onOpen() {
            updateUIShow();
        }

        public void onStart() {
            isUserControl = false;
            Platform.runLater(new Runnable() {
                public void run() {
                    homeView.updateMusicPlayState();
                }
            });
        }

        public void onStop() {

            Platform.runLater(new Runnable() {
                public void run() {
                    homeView.setButtonPlay();
                }
            });
        }

        public void onClose() {
            if (!isUserControl) {
                Platform.runLater(new Runnable() {
                    public void run() {
                        playNextAuto();
                    }
                });
            }
        }
    };


    private void playMusic() {

        MusicPlayer.getInstans().play(mPlayStateListener);

    }


    public void closePlay() {
        isUserControl = true;
        MusicPlayer.getInstans().closePlay();

    }


    private void updateUIShow() {
        Platform.runLater(new Runnable() {
            public void run() {
                int currentPlayPosition = DataManager.getInstans().getCurrentPlayPosition();
                if (currentPlayPosition != -1) {
                    homeView.listView.getSelectionModel().select(currentPlayPosition);
                }
                homeView.notifyListViewDataSetChange();

            }
        });

    }
}

File: src/com/music/read/PlayManager.kt
package com.music.read

import javafx.application.Platform

/**
 * Created by xupanpan on 03/11/2017.
 */
class PlayManager(private val homeView: HomeView) {

    private var isUserControl: Boolean = false

    var playType = PlayType.RECYCLE

    private val listener = PlayScheduleListener { position -> Platform.runLater { homeView.playTimeLabel.text = Utils.getMusicTime(position) } }


    private val mPlayStateListener = object : PlayStateListener {

        override fun onOpen() {
            updateUIShow()
        }

        override fun onStart() {
            isUserControl = false
            Platform.runLater {
                val currentPlayInfo = DataManager.instans.currentPlayInfo
                homeView.setButtonStop()
                homeView.setCurrentPlayTitle(DataManager.instans.currentPlayPosition + 1 + " / " + DataManager.instans.listSize + " # " + currentPlayInfo!!.fileName + "   " + currentPlayInfo!!.filePath)
            }
        }

        override fun onStop() {

            Platform.runLater { homeView.setButtonPlay() }
        }

        override fun onClose() {
            if (!isUserControl) {
                Platform.runLater { playNextAuto() }
            }
        }
    }

    enum class PlayType {
        SINGLE, RECYCLE, RADOM
    }


    init {
        MusicPlayer.instans.setPlayScheduleListener(listener)
    }

    fun play(position: Int) {
        play(DataManager.instans.getMP3InfoByPosition(position))
    }


    fun play(info: MP3Info?) {
        if (info == null || info.musicFile == null) {
            return
        }

        DataManager.instans.setNewPlayPosition(info)
        isUserControl = true
        playMusic()
    }

    fun playNextAuto() {

        when (playType) {
            PlayManager.PlayType.RADOM -> DataManager.instans.setPlayNextRadomPosition()
            PlayManager.PlayType.RECYCLE -> DataManager.instans.setPlayNextPosition()
            PlayManager.PlayType.SINGLE -> {
            }
        }

        playMusic()
        homeView.scrollToShow()
    }

    fun playNext() {
        isUserControl = true
        playNextAuto()
    }


    fun playLast() {
        isUserControl = true
        DataManager.instans.setPlayLastPosition()
        playMusic()
        homeView.scrollToShow()
    }

    fun stopAndStart() {

        if (MusicPlayer.instans.isPlaying) {
            isUserControl = true
            MusicPlayer.instans.closePlay()

        } else {
            playMusic()
            homeView.setButtonPlay()
            homeView.scrollToShow()
        }
    }


    private fun playMusic() {

        MusicPlayer.instans.play(mPlayStateListener)

    }


    fun closePlay() {
        isUserControl = true
        MusicPlayer.instans.closePlay()

    }


    private fun updateUIShow() {
        Platform.runLater {
            homeView.listView.setItems(null)
            homeView.listView.setItems(DataManager.instans.list)
            val currentPlayPosition = DataManager.instans.currentPlayPosition
            if (currentPlayPosition != -1) {
                homeView.listView.selectionModel.select(currentPlayPosition)
            }
        }

    }
}
